### PR TITLE
EN-1975 Validate public method arguments

### PR DESCRIPTION
**Pull request by igor-alekseev-dev**
- make country code first argument for all calls
- remove country code from record object
- simplify call signature for `delete` and `read` - the second argument is `key` string and not an object 
- remove `_` and `Async` from method's names

Pull Request Checklist
======================

Please read this for our PR Culture:
------------------------------------
https://incountry.atlassian.net/wiki/spaces/ED/pages/46858869/Pull+requests+culture+Draft

Work relates to:
----------------
Link(s) to Jira Task/User Story/Bug/Epic:
- [x] https://incountry.atlassian.net/browse/EN-1975

Code Quality (Before submitting PR)
------------
- [x] Ensure code compiles and passes Sonar Lint.
- [x] Do your best to ensure integration tests pass. If they don't then...
- [x] Remove, ignore, or mark pending any brittle tests.
- [x] Remove commented out code.

Target Release
--------------
- [x] Please put the target release vehicle this needs to be bundled with.
- [x] Did you increment the version in package.json? What is the new version?

### CODE REVIEW

File: storage.js [hunk 1]
```diff
@@ -123,34 +124,25 @@ class Storage {
...
-      country: countrycode,
-      key: record.key,
-    };
+  /**
+ * @param {string} countryCode - Country code.
+ * @param {Record} record
+ * @return {Promise<{ record: Record }>} Matching record.
+ */
+  async writeAsync(countryCode, record) {
+    if (typeof countryCode !== 'string') {
```
Comment (artemgolozubov) [inline storage.js]: Looks like this check isn't equivalent to the previous. Consider `countryCode = ''`, what will happen in this case?

File: tests/integration/update-record-tests.js [hunk 2]
```diff
@@ -147,7 +144,7 @@ describe('Update record', function () {
           body: JSON.stringify({ UpdatedName: 'UpdatedPersonName' }),
         };
 
-        await expect(storage.updateOne(data.country, { key: `NotExistingKey${Math.random().toString(36).substr(2, 10)}` }, updatedData, { override: true }))
+        await expect(storage.updateOne(COUNTRY, { key: `NotExistingKey${Math.random().toString(36).substr(2, 10)}` }, updatedData, { override: true }))
```
Comment (artemgolozubov) [inline tests/integration/update-record-tests.js]: Maybe moving this key to a separate variable as in other files above will increase readability.

File: storage.js [hunk 3]
```diff
@@ -277,22 +278,29 @@ class Storage {
...
+  /**
+   * @param {string} countryCode - Country code.
+   * @param {string} recordKey
+   * @return {Promise<{ record: Record|null }>} Matching record.
+   */
+  async readAsync(countryCode, recordKey) {
+    this._validateCountryCode(countryCode);
+    if (typeof recordKey !== 'string') {
+      this._logAndThrowError('Missing key');
+    }
```
Comment (RommeroInCountry) [inline storage.js]: should it also be smth like `_validateRecordKey`? This way it'll look more consistent

Comment (RommeroInCountry) [inline storage.js]: should it also be smth like _validateRecordKey? This way it'll look more consistent

File: storage.js [hunk 4]
```diff
@@ -7,12 +7,15 @@ const defaultLogger = require('./logger');
 const CountriesCache = require('./countries-cache');
 const SecretKeyAccessor = require('./secret-key-accessor');
 const { InCrypt } = require('./in-crypt');
-const { PositiveInt } = require('./utils');
 const {
-  StorageClientError,
   StorageServerError,
 } = require('./errors');
 
+const { validateRecord } = require('./validation/record');
```
[the remark this answers is not in the document]
Comment (igor-alekseev-dev) [reply storage.js]: 1. I am OK with separate "require" statements. It's easier to support. If storage needs another validation we will just create file and import. But with "index.js" we would import and then export. I think it's unnecessary.

2. Maybe :)

File: storage.js [hunk 5]
```diff
@@ -222,30 +226,21 @@ class Storage {
...
-      }
-
-      if (options.limit > Storage.MAX_LIMIT) {
-        this._logAndThrowError(
-          `Max limit is ${Storage.MAX_LIMIT}. Use offset to populate more`,
-        );
+  async find(countryCode, filter, options = {}) {
+    try {
+      validateCountryCode(countryCode);
+      if (options.limit) {
```
[the remark this answers is not in the document]
Comment (igor-alekseev-dev) [reply storage.js]: "Limit" is optional for find method options but when it is there it should be valid.
I would change check to `!== undefined`

File: storage.js [hunk 6]
```diff
@@ -123,34 +122,31 @@ class Storage {
     throw new Error(errorMessage);
   }
 
-  _validateRecord(record) {
-    if (!record.country) throw new StorageClientError('Missing country');
-    if (!record.key) throw new StorageClientError('Missing key');
+  _logAndReThrowError(error) {
+    this._logger.write('error', error.message);
+    throw error;
```
[the remark this answers is not in the document]
Comment (igor-alekseev-dev) [reply storage.js]: `_logAndReThrowError` rethrows whatever was passed
Actually I don't like the whole scheme of throwing errors during validation. 
But I think we need to live with it for a while.

I believe that SDK as a third-party code in client's apps shouldn't throw exceptions every time and crash everything.

File: storage.js [hunk 7]
```diff
@@ -167,13 +163,15 @@ class Storage {
    */
   async batchWrite(countryCode, records) {
     try {
+      validateCountryCode(countryCode);
+
       if (!records.length) {
         throw new Error('You must pass non-empty array');
       }
```
Comment (RommeroInCountry) [inline storage.js]: this should be a part of validation too

File: storage.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (RommeroInCountry) [reply storage.js]: Yep, let's move it inside validateLimit. Can't see why not

File: tests/unit/storage-tests.js [hunk 8]
```diff
@@ -708,7 +731,7 @@ describe('Storage', () => {
       nock.enableNetConnect();
     });
 
-    describe('_getEndpointAsync', () => {
+    describe('_getEndpoint', () => {
```
Comment (artemgolozubov) [inline tests/unit/storage-tests.js]: Do we still need underscore here?

**Pull request merged**